# Module 10: Descriptive Statistics and Outliers
***

## What are statistics?

Statistics deal with collecting informative data, interpreting these data, and drawing conclusions about the data. Statistics are used across multiple disciplines and fuel business and policy decisions.

## Two types of statistics
***
#### Descriptive Statistics
Descriptive statistics use numeric and graphical techniques to summarize the characteristics of a set of data and to identify patterns within the dataset. Descriptive statistics also include graphically summarizing data and presenting the information in a meaningful (easy to interpret) way. 

#### Inferential Statistics
Inferential statistics use samples of data to make generalizations about a larger set of data (or a larger population). Inferential statistics use data to make estimates, decisions, and predictions about patterns and trends identified within a set of data. Inferential statistics are used to make conclusions beyond your immediate set of data. 

## Import Libraries and Data

In [4]:
import pandas as pd
import numpy as np
import scipy.stats as stats ## new library alert! ##

## SciPy is used for scientific and mathematic computing; it provides functions for stats, 
# data manipulation, and data visualization

df = pd.read_excel("axisdata.xlsx")

df.head()

Fname      Lname Position Gender  Hours Worked SalesTraining  \
0  Jackie    Jackson  Trainee      F            32             Y   
1    Mary  Patterson  Trainee      F            43             N   
2   Tanya      Adams  Trainee      F            28             Y   
3   Tanya  Henderson  Trainee      F            24             Y   
4  Walter   Franklin  Trainee      M            25             Y   

   Years Experience  Cars Sold  
0                 1          3  
1                 1          6  
2                 1          2  
3                 1          5  
4                 1          4

## Qualitative vs Quantitative Data
***

#### Quantitative Data
Data that are recorded on a numeric scale (i.e. age, weight, salary, number of empty seats on a plane, credit score etc). Quantitative variables are also called <b>numeric</b> variables. 

#### Qualitative Data
Data that cannot be measured on a numeric scale (i.e. eye color, type of car, blood type, race etc). These data can only be classified into one group of categories. Qualitative variables are also called <b>categorical</b> variables. Qualitative data also includes text and string data. 

In [5]:
## Which variables are qualitative and which variables are quantitative?

df.head()

Fname      Lname Position Gender  Hours Worked SalesTraining  \
0  Jackie    Jackson  Trainee      F            32             Y   
1    Mary  Patterson  Trainee      F            43             N   
2   Tanya      Adams  Trainee      F            28             Y   
3   Tanya  Henderson  Trainee      F            24             Y   
4  Walter   Franklin  Trainee      M            25             Y   

   Years Experience  Cars Sold  
0                 1          3  
1                 1          6  
2                 1          2  
3                 1          5  
4                 1          4

## Describing Qualitative Data
***

Qualitative data is best summarized by frequency tables. 

Frequency tables summarize categorical data into a table showing each category, number of observations that fall within each category, and the relative frequency of each category. 

Relative frequency is the number of specific observations out of the total number of observations, or the percentage of the total sample that fall within a specific group. Both representations of categorical data give insight into how common specific categories are within a set of data. 

<b>Bar charts</b> are ideal for graphically representing qualitative data as they visually show the frequency of observations across various groups. 

#### GRAPHIC SHOULD READ N = 150 !!!

<img src="FreqTable.png">

In [6]:
## Frequencies 

df["Gender"].value_counts()

M    510
F    489
Name: Gender, dtype: int64

In [7]:
## Relative Frequencies 

df["Gender"].value_counts(normalize=True)

M    0.510511
F    0.489489
Name: Gender, dtype: float64

In [8]:
## Multi-variable Frequencies #normalize a relative frequency is % score
## True is a percentage

pd.crosstab(df["Gender"], df["SalesTraining"], margins=False, normalize=False)

SalesTraining    N    Y
Gender                 
F              212  277
M              201  309

# { Exercise 1 }

    1. Import the "pokemon.csv" file; name the dataset 'poke'. Preview the first 5 rows. 
    2. Determine the frequencies of unique pokemon types (variable name: Type 1) in this dataset
    3. Determine the relative frequencies of the unique pokemon types (variable name: Type 1) in this dataset
    4. Create a crosstabs table that shows the frequency of pokemon stage (variable name: Stage) and legendary status (variable name: Legendary). 

In [9]:
#1-1
poke = pd.read_csv("pokemon.csv")

df.head()

Fname      Lname Position Gender  Hours Worked SalesTraining  \
0  Jackie    Jackson  Trainee      F            32             Y   
1    Mary  Patterson  Trainee      F            43             N   
2   Tanya      Adams  Trainee      F            28             Y   
3   Tanya  Henderson  Trainee      F            24             Y   
4  Walter   Franklin  Trainee      M            25             Y   

   Years Experience  Cars Sold  
0                 1          3  
1                 1          6  
2                 1          2  
3                 1          5  
4                 1          4

In [10]:
#1-2
poke["Type 1"].value_counts()

Water       28
Normal      22
Poison      14
Grass       12
Fire        12
Bug         12
Electric     9
Rock         9
Ground       8
Psychic      8
Fighting     7
Ghost        3
Dragon       3
Fairy        2
Ice          2
Name: Type 1, dtype: int64

In [11]:
#1-3
poke["Type 1"].value_counts(normalize = True)

Water       0.185430
Normal      0.145695
Poison      0.092715
Grass       0.079470
Fire        0.079470
Bug         0.079470
Electric    0.059603
Rock        0.059603
Ground      0.052980
Psychic     0.052980
Fighting    0.046358
Ghost       0.019868
Dragon      0.019868
Fairy       0.013245
Ice         0.013245
Name: Type 1, dtype: float64

In [12]:
#1-4
pd.crosstab(poke["Legendary"], poke["Stage"], margins=True, normalize=False)

Stage       1   2   3  All
Legendary                 
False      75  56  16  147
True        4   0   0    4
All        79  56  16  151

## Describing Quantitative Data

There are many more techniques for describing quantitative data compared to qualitative. When summarizing quantitative data, the objective is to get a sense of how the data is clustered and the variation in numeric values. 

***

### Measures of Central Tendency

Measures of Central Tendency are estimates of what a "typical" value/data point is within your set of data. A measure of central tendency is a number around which an entire sample of data is spread. We can describe this central position using several statistics, including the mean (average), the median, and the mode. 

Measures of central tendency can be visualized with a frequency distribution (histogram). A frequency distribution is a description of values and how often each value occurs within your dataset. When you graph a frequency distribution a pattern/shape should give insight into which values occured the most/least. 

<img src="NormalDist.jpg">

***

#### Mean

The mean, which is the most commonly known and used measure of central tendency, is the arithmetic average of a set of numeric values. The mean is calculated by adding all the values together and dividing by the total number of values in the set of values. In specific sets of data, the mean is the central point in the frequency distribution aka the hump. 

In [13]:
## Calculate the Mean of a column 

df["Hours Worked"].mean()

33.727727727727725

#### Median 

The median is the value which divides the data into two equal parts when the data is arranged from least to greatest. The median will be the middle value if the number of values is odd, otherwise, it will be the average of the two middle values if the number of values is even. 

In [14]:
## Calculate the Median of a column 

df["Hours Worked"].median()

34.0

#### Mode 

The mode is the value that appears most frequently in a sample of data, or the value that has the highest frequency. It is possible for a set of data to have no mode, and it is possible for a set of data to have multiple modes. 

In [15]:
## Calculate the Mode of a column 

df["Hours Worked"].mode()

0    38
Name: Hours Worked, dtype: int64

# { Exercise 2 }

    1. Using the pokemon dataset ('poke'), calculate the mean attack value (Attack) in the dataset.  
    2. Determine the median defense value (Defense). Write a single sentence on what your findings mean. 
    3. Determine the mode of the speed (Speed) column. Write a single sentence on what your findings mean. 

In [19]:
#2-1
df["Hours Worked"].mean()

33.727727727727725

In [20]:
#2-2
poke["Attack"].mean()


72.54966887417218

In [21]:
poke["Defense"].median()

65.0

In [22]:
poke["Speed"].mode()

0    70
1    90
Name: Speed, dtype: int64

### Data Distribution and Measurement Sensitivity
***

### Normally distributed data vs. non-normally distributed data

Several statistics are suited for data that follows a <b>normal distribution</b>. When visualized, normally distributed data lies in a symmetrical pattern (bell-shape) where the majority of the data surrounds the mean with decreasing amounts evenly distributed to the left and the right. 

While this is the ideal, real-world data is not always normally distributed. Non-normal distribution is possible and there are several statistics that can still be used with non-normally distributed data. It is important to determine how your data is distributed before you embark on certain statistics. 

#### Distribution and Measure of Central Tendency

The shape of the data helps to determine which measure of central tendency should be reported. Some statistics are very sensitive to non-normal distributions, while others are more resilient. Three of the most common shapes to learn are symmetric, left-skewed, and right-skewed. <b>Skew</b> is simply a measure of the asymmetry of the distribution. <b>In the below images, while not shown, the values are increasing from left to right.</b> 

#### Symmetric Data
* mean, median, and mode are all the same here
* no skewness is apparent
* the distribution is described as symmetric

<img src="Symmetrical.jpg">

***

#### Left-skewed Data
* mean < median
* long tail on the left

<img src="LeftSkew.jpg">

***

#### Right-skewed Data
* mean > median
* long tail on the right

<img src="RightSkew.jpg">

***

### Considerations for Shape

When your data is skewed in either direction, you should consider the best measure of central tendency to report. The median is best used when you have skewed data. This is because the median is less affected by extreme values. 

If your data is skewed, this could be an indication that your data has extreme high or low values that are contributing to the skewness. Later in this lesson, we will talk about how to handle extreme values. 

### Measures of Variability (or Spread)

Measures of Variability, or spread, describe the amount of dispersion in your data sample. In other words, measures of variability, describe how spread out values are. For example, in a sample of 100 students, the average test score is 75. However, not all students will have gotten a 75 - their scores will be spread out above and below the average, some will be higher and some lower. To describe the spread of our data, we can use several statistics, including range, standard deviation, percentiles, and quartiles.

***

#### Range

The range is the difference between the lowest and highest value in a set of data. For example, if the maximum value is 10 and the minimum value is 2, the range is 8. The larger the range, the larger the difference between the highest and lowest values. 

In [23]:
## Calculate the range of values in a column

hrs_range = df['Hours Worked'].max() - df['Hours Worked'].min()

print(hrs_range)

28


#### Standard Deviation (SD)

The SD is a measurement of the average distance between each data value and the mean. You can consider SD as a measure of how spread out the data is from the mean. A low SD indicates that the data points are clustered around the mean, while a high SD indicates the data points are spread out over a wider range of values. 

<img src="https://d20khd7ddkh5ls.cloudfront.net/high_low_standard_deviation.png">

In [24]:
## Calculate the Standard Deviation of values in a column

print("Mean hours worked by all employees:", df["Hours Worked"].mean())

df["Hours Worked"].std()

Mean hours worked by all employees: 33.727727727727725


8.223453795492198

#### Percentiles

Percentiles give insight into the relation of one data point to other data points within the same dataset. When the data values are arranged from lowest to highest, percentiles represent the position of a value within the list of values. For example, if a value is at the 50th percentile - 50% of the data falls below the value, and 50% fall above the value. If a value is at the 85th percentile - 85% of the data falls below the value, and 15% fall above the value. 

In [25]:
## Calculate percentiles of values in a column
## To determine the value that falls at a specific percentile
## np.percentile(data/column, percentile value)

np.percentile(df['Hours Worked'], 85)

## Determine which value falls at the 85th percentile? 
## 85% of the values in Hours Worked are below this value, 15% are above this value. 

44.0

#### Quartiles 

Quartiles are values that divide a set of data into equal quarters (when the data is ordered from lowest to highest). There are <b>three</b> quartile values that divide the data into four parts (see below). 

<img src="quar.png">

* The first quartile (Q1) is at the 25th percentile (25% below, 75% above)
* The second quartile (Q2) is at the 50th percentile (50% below, 50% above, median)
* The third quartile (Q3) is at the 75th percentile (75% below, 25% above)

<b>Quartiles can be used to calculate the Interquartile Range (IQR) - which can be used to handle outliers in your dataset.</b>

In [26]:
## Determine the three quartiles of a set of data
## A quantile is a series of values, or cut points, that divide a set of data into equal parts 
## A 'quartile' is a specical kind of quantile that divides the data into four equal parts

df['Hours Worked'].quantile(.25) 

#Q1 = .25
#Q2 = .50
#Q3 = .75

27.0

#### All of the Above 

The describe function is perfect for outputting multiple descriptive statistics at once. This will save you multiple steps if you want to look at one or several quantitative variables. 

In [27]:
## Use the describe function to view multiple statistics at the same time

df.describe()

Hours Worked  Years Experience   Cars Sold
count    999.000000        999.000000  999.000000
mean      33.727728          3.026026    3.922923
std        8.223454          1.394709    1.527000
min       20.000000          1.000000    1.000000
25%       27.000000          2.000000    3.000000
50%       34.000000          3.000000    4.000000
75%       41.000000          4.000000    5.000000
max       48.000000          5.000000    7.000000

In [28]:
## Use the describe function to isolate information for a single column

df['Hours Worked'].describe()

count    999.000000
mean      33.727728
std        8.223454
min       20.000000
25%       27.000000
50%       34.000000
75%       41.000000
max       48.000000
Name: Hours Worked, dtype: float64

# { Exercise 3 }

    1. Using the pokemon dataset ('poke'), calculate the range for the HP column. 
    2. What is the mean and the standard deviation for the special attack (SpAtk) column?
    3. For the Attack column, determine which value falls at the 80th percentile? Which value falls at the 15th percentile? For each of these values, write a single sentence describing what this means. 
    4. Calculate the second quartile (or the median) for the special defense (SpDef) column. Then, calculate the median of the same column using the method shown above. Are the values the same? 

In [29]:
#3-1
poke


Num        Name   Type 1  Type 2  Total   HP  Attack  Defense  SpAtk  \
0      1   Bulbasaur    Grass  Poison    318   45      49       49     65   
1      2     Ivysaur    Grass  Poison    405   60      62       63     80   
2      3    Venusaur    Grass  Poison    525   80      82       83    100   
3      4  Charmander     Fire     NaN    309   39      52       43     60   
4      5  Charmeleon     Fire     NaN    405   58      64       58     80   
..   ...         ...      ...     ...    ...  ...     ...      ...    ...   
146  147     Dratini   Dragon     NaN    300   41      64       45     50   
147  148   Dragonair   Dragon     NaN    420   61      84       65     70   
148  149   Dragonite   Dragon  Flying    600   91     134       95    100   
149  150      Mewtwo  Psychic     NaN    680  106     110       90    154   
150  151         Mew  Psychic     NaN    600  100     100      100    100   

     SpDef  Speed  Stage  Legendary  
0       65     45      1      False  
1       80     60      2      False  
2      100     80      3      False  
3       50     65      1      False  
4       65     80      2      False  
..     ...    ...    ...        ...  
146     50     50      1      False  
147     70     70      2      False  
148    100     80      3      False  
149     90    130      1       True  
150    100    100      1      False  

[151 rows x 13 columns]

In [30]:

poke_range = poke["HP"].max() - poke["HP"].min()
print(poke_range)

240


In [31]:
print("Mean special for pokemon is:", poke["SpAtk"].mean())

poke["SpAtk"].std()

      


Mean special for pokemon is: 67.13907284768212


28.53419930191353

## Handling and Removing Outliers

***

Outliers are data points that differ significantly from other data points. Outliers are an important feature of a dataset that should be identified and addressed because extreme values can influence the results of your statistics. 

   * In a sample of 5 students, their ages are 25, 22, 97, 23, 21 - 97 is an outlier because the value differs significantly from the remaining data.
   * Without removing outlier, the mean is: 37.6 and the median is: 23
   * When removing the outlier, the mean is: 22.75 and the median is: 22.5

Determining which values are extreme can be challenging when you have a larger dataset. There are several methods that can be used to strategically determine which values are outliers, we will cover two separate methods below. 

In [32]:
df = pd.read_csv("SalaryData.csv")
df.head()

Employee  YearsExperience   Salary  Age  Gender  RemoteStatus   Region  \
0  Richard              5.1    66029   41  Female             0  Central   
1     Nick              5.9  8136300   39  Female             0  Central   
2   Morgan              9.5   116969   50  Female             1  Central   
3    Susan              3.2    64445   27    Male             0  Central   
4  Matthew              8.2   113812   47  Female             1  Central   

  CompanyCar  Shift  PerformanceReviewScore  
0          Y      2                    98.3  
1          Y      3                   105.7  
2          N      1                   145.0  
3          Y      3                    75.1  
4          N      2                   131.1

## Method 1 : Using z-scores to detect and remove outliers

In statistics, the <font color=salmon><b>Empirical Rule</b></font> estimates how data, following a normal distribution, will spread around the mean. According to this rule, basically all data values will fall within 3 SD's above or below the mean. The Empirical Rule is also known as the <font color=salmon><b>68-95-99.7 Rule</b></font> because according to this rule, your data will typically fall into the following pattern:

* 68% of the data falls within one SD of the mean

* 95% of the data falls within two SD's of the mean

* 99.7% of the data falls within three SD's of the mean

<img src="SD Bell Curve.jpeg">

#### Empirical Rule Example:

The average weight of an American man is 175 lbs with a SD of 10 lbs. According to the Empirical Rule, we can assume: 

* 68% of American men are between 165 and 185 lbs (or +/- 10 lbs from the mean, or 1 SD)
* 95% of American men are between 155 and 195 lbs (or +/- 20 lbs from the mean, or 2 SD's)
* 99.7% of American men are between 145 and 205 lbs (or +/- 30 lbs from the mean, or 3 SD's)
* Typically, data values that fall outside of this range are considered outliers
_______________________ 
A <font color=salmon><b>z-score</b></font>, also known as a standard score, measures the distance of a value from the mean. In other words, the z-score tells you how many SD's a given value is from the mean. Values that fall above the mean will have a positive (+) z-score, while values that fall below the mean will have a negative (-) z-score. Values that are far from the mean will be considered outliers. You can define a value as an outlier if the z-score greater than +/- 3. 
_______________________ 

#### Steps for Detecting and Removing Outliers with z-scores
* Import the scipy library to make use of the z-score function
* Calculate the z-score for each value within a specific column in your dataset
* Transform the z-score to the absolute value of the z-score (number without the +/- sign; this will make our code simpler)
* Add the z-scores to your dataset as a new column
* Remove all rows that have a z-score greater than 3

In [33]:
## Create a copy of your dataset to filter outliers
dfz = df.copy()

## Check original shape of the dataset
print(dfz.shape)

(30, 10)


In [34]:
## Create a new column to contain the z-scores for the employee salary
## Set the new column equal to the absolute value of the z-scores
## To calculate z-score -- stats.zscore(data/column name)
## To determine absolute value of z-score -- np.abs(data/function/options)

dfz["zscore_salary"] = np.abs(stats.zscore(dfz["Salary"]))

## Preview new column; optional
dfz.head()

Employee  YearsExperience   Salary  Age  Gender  RemoteStatus   Region  \
0  Richard              5.1    66029   41  Female             0  Central   
1     Nick              5.9  8136300   39  Female             0  Central   
2   Morgan              9.5   116969   50  Female             1  Central   
3    Susan              3.2    64445   27    Male             0  Central   
4  Matthew              8.2   113812   47  Female             1  Central   

  CompanyCar  Shift  PerformanceReviewScore  zscore_salary  
0          Y      2                    98.3       0.269844  
1          Y      3                   105.7       3.343557  
2          N      1                   145.0       0.247036  
3          Y      3                    75.1       0.270553  
4          N      2                   131.1       0.248449

In [35]:
## Determine the index locations for the rows with zscores that are greater than "3"
z_outliers = dfz.loc[dfz["zscore_salary"] > 3].index

## Preview list of index values
print(z_outliers)

Int64Index([1, 28], dtype='int64')


In [36]:
## What information can we find at these index locations?

dfz.iloc[[1, 28]]

Employee  YearsExperience   Salary  Age  Gender  RemoteStatus   Region  \
1      Nick              5.9  8136300   39  Female             0  Central   
28     Bill              7.1  9827300   36  Female             0     East   

   CompanyCar  Shift  PerformanceReviewScore  zscore_salary  
1           Y      3                   105.7       3.343557  
28          Y      3                   116.8       4.100689

In [37]:
## Drop rows with above index values
dfz = dfz.drop(z_outliers)

## Re-check the shape of the dataframe, how many rows were dropped?
print(dfz.shape)

(28, 11)


## Method 2 : Using the Interquartile Range (IQR) to detect and remove outliers

The <font color=salmon><b>Interquartile Range (IQR)</b></font> is the difference between the first and third quartile in a set of data. The IQR can be considered where the bulk of the data values rest. IQR is calculated by the following equation: <B>IQR = Q3 - Q1</B>
<img src="IQR.png">
_____________________

The IQR can be used to create "fences" or cut-off values that determine which values fall outside of the acceptable data range. For example, we can calculate a lower-limit -- all values falling below this limit will be considered an outlier. The <b>1.5 Rule</b> is a commonly used method for determining the upper and lower cut-off limits. According to this rule, a value is considered an outlier if it falls:

* ... below Q1 - (1.5 x IQR)

* ... above Q3 + (1.5 x IQR)
_____________________

#### 1.5 Rule Example:

A professor is tracking the number of absences for each of their students. They have compiled a list of the number of absences: 

    1, 3, [4], 6, 7, [7], 8, 8, [9], 22, 37 

In this list of data, we can find the following values:

* Q1 = 4
* Q2 (median) = 7
* Q3 = 9

With these values, you can easily calculate the IQR:

* <B>IQR</B> = Q3(9) - Q1(4)
* <B>IQR</B> = 5

Once you have the IQR, you can calculate the upper and lower limits for detecting outliers:

* Upper Limit = Q3(9) + [1.5 * IQR(5)] = <b> 16.5 </b>
* Lower Limit = Q1(4) - [1.5 * IQR(5)] = <b> -3.5 </b>

Values that fall above the upper limit are outliers; values that fall below the lower limit are outliers
_______________________ 

#### Steps for Detecting and Removing Outliers with IQR

* Calculate quartiles
* Calculate IQR
* Determine upper and lower limits
* Drop values less than the lower and/or greater than the upper

In [38]:
## Create a copy of original dataset
dfq = df.copy()

## Check original shape of the dataset
dfq.shape

(30, 10)

In [39]:
## Calculate quartiles
q1 = dfq["Salary"].quantile(.25)
q3 = dfq["Salary"].quantile(.75)

print("Q1:", q1)
print("Q3:", q3)

Q1: 48265.0
Q3: 104512.0


In [40]:
dfq

Employee  YearsExperience   Salary  Age  Gender  RemoteStatus   Region  \
0   Richard              5.1    66029   41  Female             0  Central   
1      Nick              5.9  8136300   39  Female             0  Central   
2    Morgan              9.5   116969   50  Female             1  Central   
3     Susan              3.2    64445   27    Male             0  Central   
4   Matthew              8.2   113812   47  Female             1  Central   
5   Richard             10.3      160   47    Male             1  Central   
6     James              3.0    60150   43    Male             1     West   
7     Smith              3.9    63218   49  Female             1     West   
8      Bill             10.5   121872   44    Male             1     West   
9   Richard              2.2    39891   40    Male             0  Central   
10    James              5.3    83088   40  Female             0  Central   
11   Morgan              8.7     1094   27  Female             0  Central   
12   Thomas              1.5   377310   46    Male             1  Central   
13  Richard              4.0    55794   48    Male             1  Central   
14   Rachel              1.3    46205   32  Female             0     East   
15    Susan              4.5     6111   50    Male             1  Central   
16     Alex              6.8    91738   31  Female             0  Central   
17  Matthew              1.1    39343   38  Female             0     East   
18     Alex              2.9    56642   37  Female             0     East   
19    Smith              3.7      300   42  Female             0     East   
20   Rachel              4.1    57081   27    Male             0     East   
21    Susan              7.9   101302   29    Male             0     East   
22  Richard              9.6   112635   48  Female             1     East   
23     Bill              3.2    54445   38    Male             1  Central   
24  Matthew              6.0    93940   33    Male             0  Central   
25    Smith              9.0   105582   28  Female             0  Central   
26     Alex              2.0    43525   34    Male             0     East   
27  Richard              4.9    67938   40  Female             0     East   
28     Bill              7.1  9827300   36  Female             0     East   
29    James              4.0    56957   45    Male             1     West   

   CompanyCar  Shift  PerformanceReviewScore  
0           Y      2                    98.3  
1           Y      3                   105.7  
2           N      1                   145.0  
3           Y      3                    75.1  
4           N      2                   131.1  
5           N      3                   152.1  
6           Y      1                    77.9  
7           N      3                    88.7  
8           N      3                   153.2  
9           Y      1                    69.0  
10          Y      2                   100.0  
11          N      1                   130.1  
12          Y      1                    63.8  
13          Y      2                    89.4  
14          Y      1                    57.6  
15          Y      3                    95.0  
16          Y      1                   112.3  
17          Y      1                    57.4  
18          Y      3                    75.1  
19          Y      3                    84.6  
20          Y      2                    84.1  
21          Y      2                   122.7  
22          N      2                   145.4  
23          Y      1                    78.4  
24          Y      1                   104.9  
25          N      3                   133.4  
26          Y      3                    65.2  
27          Y      3                    96.0  
28          Y      3                   116.8  
29          N      2                    88.5

In [41]:
## Calculate the IQR
iqr = q3 - q1

print("IQR:", iqr)

IQR: 56247.0


In [42]:
## Determine outlier fences 
top = q3 + (iqr * 1.5)
bottom = q1 - (iqr * 1.5)


print("Upper Limit:", top)
print("Lower Limit:", bottom)

Upper Limit: 188882.5
Lower Limit: -36105.5


In [43]:
## Determine the index locations for rows that fall outside of outlier fences

iqr_outliers = dfq.loc[(dfq['Salary'] > top) | (dfq['Salary'] < bottom)].index

print("INDEX VALUES:", iqr_outliers)

INDEX VALUES: Int64Index([1, 12, 28], dtype='int64')


In [44]:
## what values can we find at these index locations?

dfq.iloc[[1, 12, 28]]

Employee  YearsExperience   Salary  Age  Gender  RemoteStatus   Region  \
1      Nick              5.9  8136300   39  Female             0  Central   
12   Thomas              1.5   377310   46    Male             1  Central   
28     Bill              7.1  9827300   36  Female             0     East   

   CompanyCar  Shift  PerformanceReviewScore  
1           Y      3                   105.7  
12          Y      1                    63.8  
28          Y      3                   116.8

In [45]:
## Drop rows with above index values
dfq = dfq.drop(iqr_outliers)

## Re-check the shape of the dataframe, how many rows were dropped?
print(dfq.shape)

(27, 10)


# { Module 10 Homework }

1. Import the "babies.xlsx" dataset. See below for information on the columns:

    * bwt - birth weight of newborn baby
    * gestation	- gestation length (weeks)
    * parity - previously pregnant (0 = no; 1 = yes)
    * age - age of mother
    * height - height of mother (inches)	
    * weight - weight of mother (pounds)
    * smoke - smoking status of mother (0 = nonsmoker; 1 = smoker)

In [46]:
baby = pd.read_excel("babies.xlsx")

2. Preview the first few rows of the dataset. Complete the following checks for your dataset:

    * output the summary information for the dataset, what types of variables make up this dataset?
    * is there any missing data in this dataset?

In [47]:
baby.head()

bwt  gestation  parity   age  height  weight  smoke
0  120      284.0       0  27.0    62.0   100.0    0.0
1  113      282.0       0  33.0    64.0   135.0    0.0
2  128      279.0       0  28.0    64.0    77.0    1.0
3  123        NaN       0  36.0    69.0   190.0    0.0
4  108      282.0       0  23.0    67.0   125.0    1.0

3. Handle the missing data in the dataset -- there isn't much, so we can drop all the rows that have at least one missing value. How many rows were dropped from the dataset?

In [48]:
baby.dropna(inplace = True)
baby.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1174 entries, 0 to 1235
Data columns (total 7 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   bwt        1174 non-null   int64  
 1   gestation  1174 non-null   float64
 2   parity     1174 non-null   int64  
 3   age        1174 non-null   float64
 4   height     1174 non-null   float64
 5   weight     1174 non-null   float64
 6   smoke      1174 non-null   float64
dtypes: float64(5), int64(2)
memory usage: 73.4 KB


4. There are two qualitative variables in this dataset. What are they? How do you know they are qualitative variables?

In [44]:
1236-1174  # smoke and parity are the two qualitative values. They are numeric vairables of T/F. 

62

5. Replace the values in the qualitative variables with meaningful labels that describe the different groups. Use any method you've learned in previous modules to complete this task. 

In [49]:
baby["smoke"].replace([1,0],["Yes","No"], inplace=True)
baby


bwt  gestation  parity   age  height  weight smoke
0     120      284.0       0  27.0    62.0   100.0    No
1     113      282.0       0  33.0    64.0   135.0    No
2     128      279.0       0  28.0    64.0    77.0   Yes
4     108      282.0       0  23.0    67.0   125.0   Yes
5     136      286.0       0  25.0    62.0    93.0    No
...   ...        ...     ...   ...     ...     ...   ...
1231  113      275.0       1  27.0    60.0   100.0    No
1232  128      265.0       0  24.0    67.0   120.0    No
1233  130      291.0       0  30.0    65.0    67.0   Yes
1234  125      281.0       1  21.0    65.0   110.0    No
1235  117      297.0       0  38.0    65.0   129.0    No

[1174 rows x 7 columns]

In [50]:
baby.head()

bwt  gestation  parity   age  height  weight smoke
0  120      284.0       0  27.0    62.0   100.0    No
1  113      282.0       0  33.0    64.0   135.0    No
2  128      279.0       0  28.0    64.0    77.0   Yes
4  108      282.0       0  23.0    67.0   125.0   Yes
5  136      286.0       0  25.0    62.0    93.0    No

6. Before we move forward with any statistics, let's identify and remove any outliers from the dataset. Using the IQR method, search for outliers in the 5 numeric variables. This will take some time and organization, be careful with your code! Make sure you keep track of how many rows/outliers are removed. 

In [47]:
#6-1 

#Q1 = .25
#Q2 = .50
#Q3 = .75


In [51]:

baby['bwt'].quantile(.25)




108.0

In [52]:
baby['bwt'].quantile(.75)

131.0

In [53]:
q1 = baby["bwt"].quantile(.25)
q3 = baby["bwt"].quantile(.75)
print("Q1:", q1)
print("Q3:", q3)


Q1: 108.0
Q3: 131.0


In [54]:
baby

bwt  gestation  parity   age  height  weight smoke
0     120      284.0       0  27.0    62.0   100.0    No
1     113      282.0       0  33.0    64.0   135.0    No
2     128      279.0       0  28.0    64.0    77.0   Yes
4     108      282.0       0  23.0    67.0   125.0   Yes
5     136      286.0       0  25.0    62.0    93.0    No
...   ...        ...     ...   ...     ...     ...   ...
1231  113      275.0       1  27.0    60.0   100.0    No
1232  128      265.0       0  24.0    67.0   120.0    No
1233  130      291.0       0  30.0    65.0    67.0   Yes
1234  125      281.0       1  21.0    65.0   110.0    No
1235  117      297.0       0  38.0    65.0   129.0    No

[1174 rows x 7 columns]

In [55]:
iqr = q3 - q1

print("IQR:", iqr)

IQR: 23.0


In [56]:
top = q3 + (iqr * 1.5)
bottom = q1 - (iqr * 1.5)


print("Upper Limit:", top)
print("Lower Limit:", bottom)

Upper Limit: 165.5
Lower Limit: 73.5


In [57]:
q1 = baby["bwt"].quantile(.25)
q3 = baby["bwt"].quantile(.75)
print("Q1:", q1)
print("Q3:", q3)


Q1: 108.0
Q3: 131.0


In [58]:
baby["zscore_bwt"] = np.abs(stats.zscore(baby["bwt"]))

In [59]:
iqr_outliers = baby.loc[(baby['bwt'] > top) | (baby['bwt'] < bottom)].index

print("INDEX VALUES:", iqr_outliers)

INDEX VALUES: Int64Index([ 239,  309,  361,  462,  500,  529,  556,  594,  632,  709,  738,
             747,  829,  904,  912,  978, 1021, 1035, 1063, 1065, 1099, 1139,
            1148, 1169],
           dtype='int64')


In [60]:
baby.iloc[[239,  309,  361,  462,  500,  529,  556,  594,  632,  709,  738,
             747,  829,  904,  912,  978, 1021, 1035, 1063, 1065, 1099, 1139,
            1148, 1169]]

bwt  gestation  parity   age  height  weight smoke  zscore_bwt
261    81      256.0       0  30.0    64.0   148.0   Yes    2.099384
332   103      273.0       1  22.0    64.0   110.0   Yes    0.898567
388    98      275.0       0  25.0    65.0   112.0   Yes    1.171480
497   101      289.0       1  31.0    60.0   125.0    No    1.007732
537   147      277.0       0  30.0    68.0   160.0    No    1.503067
566   133      280.0       1  25.0    61.0   130.0    No    0.738911
593   118      297.0       0  35.0    68.0   140.0   Yes    0.079828
632   176      293.0       1  19.0    68.0   180.0    No    3.085961
676   100      275.0       0  26.0    60.0   115.0    No    1.062315
757   155      279.0       0  33.0    61.0   125.0    No    1.939727
787   125      272.0       0  30.0    60.0    96.0    No    0.302250
796   152      286.0       1  19.0    67.0   135.0    No    1.775980
881   128      272.0       1  18.0    67.0   109.0    No    0.465998
958   133      290.0       0  21.0    64.0   145.0    No    0.738911
967   119      273.0       0  35.0    65.0   125.0   Yes    0.025246
1035   65      237.0       0  31.0    67.0   130.0    No    2.972705
1079  119      286.0       1  33.0    67.0   137.0    No    0.025246
1093  108      256.0       1  26.0    67.0   130.0    No    0.625654
1121  128      277.0       0  39.0    61.0   120.0    No    0.465998
1123  133      289.0       0  22.0    65.0   123.0   Yes    0.738911
1157  141      293.0       0  28.0    61.0   125.0    No    1.175571
1200   97      255.0       1  22.0    63.0   107.0   Yes    1.226062
1209  144      283.0       1  25.0    66.0   140.0    No    1.339319
1231  113      275.0       1  27.0    60.0   100.0    No    0.352741

In [61]:
#6-2
baby['gestation'].quantile(.25)

272.0

In [62]:
baby['gestation'].quantile(.75)

288.0

In [63]:
q1 = baby['gestation'].quantile(.25)
q3 = baby['gestation'].quantile(.75)
print("Q1:", q1)
print("Q3:", q3)

Q1: 272.0
Q3: 288.0


In [64]:
baby

bwt  gestation  parity   age  height  weight smoke  zscore_bwt
0     120      284.0       0  27.0    62.0   100.0    No    0.029337
1     113      282.0       0  33.0    64.0   135.0    No    0.352741
2     128      279.0       0  28.0    64.0    77.0   Yes    0.465998
4     108      282.0       0  23.0    67.0   125.0   Yes    0.625654
5     136      286.0       0  25.0    62.0    93.0    No    0.902658
...   ...        ...     ...   ...     ...     ...   ...         ...
1231  113      275.0       1  27.0    60.0   100.0    No    0.352741
1232  128      265.0       0  24.0    67.0   120.0    No    0.465998
1233  130      291.0       0  30.0    65.0    67.0   Yes    0.575163
1234  125      281.0       1  21.0    65.0   110.0    No    0.302250
1235  117      297.0       0  38.0    65.0   129.0    No    0.134411

[1174 rows x 8 columns]

In [65]:
iqr = q3 - q1

print("IQR:", iqr)

IQR: 16.0


In [66]:
top = q3 + (iqr * 1.5)
bottom = q1 - (iqr * 1.5)


print("Upper Limit:", top)
print("Lower Limit:", bottom)

Upper Limit: 312.0
Lower Limit: 248.0


In [67]:
baby["zscore_gestation"] = np.abs(stats.zscore(baby["gestation"]))

In [68]:
iqr_outliers = baby.loc[(baby['gestation'] > top) | (baby['gestation'] < bottom)].index

print("INDEX VALUES:", iqr_outliers)

INDEX VALUES: Int64Index([   6,    7,   10,   59,   63,   66,  119,  129,  155,  188,  192,
             198,  210,  215,  217,  234,  240,  253,  260,  279,  345,  361,
             373,  394,  440,  460,  462,  484,  500,  511,  523,  630,  685,
             699,  710,  726,  746,  761,  769,  778,  784,  828,  829,  830,
             833,  869,  912,  927,  952,  969,  978, 1002, 1020, 1026, 1035,
            1065, 1074, 1134, 1139, 1140, 1142, 1146, 1152, 1172, 1178, 1199,
            1206, 1207, 1217, 1219, 1226],
           dtype='int64')


In [69]:
baby.loc[[6, 7,   10,   59,   63,   66,  119,  129,  155,  188,  192,
             198,  210,  215,  217,  234,  240,  253,  260,  279,  345,  361,
             373,  394,  440,  460,  462,  484,  500,  511,  523,  630,  685,
             699,  710,  726,  746,  761,  769,  778,  784,  828,  829,  830,
             833,  869,  912,  927,  952,  969,  978, 1002, 1020, 1026, 1035,
            1065, 1074, 1134, 1139, 1140, 1142, 1146, 1152, 1172, 1178, 1199,
            1206, 1207, 1217, 1219, 1226]]

bwt  gestation  parity   age  height  weight smoke  zscore_bwt  \
6     138      244.0       0  33.0    62.0   178.0    No    1.011823   
7     132      245.0       0  23.0    65.0   140.0    No    0.684328   
10    140      351.0       0  27.0    68.0   120.0    No    1.120989   
59     75      232.0       0  33.0    61.0   110.0    No    2.426879   
63    101      245.0       0  23.0    63.0   130.0   Yes    1.007732   
...   ...        ...     ...   ...     ...     ...   ...         ...   
1206  127      242.0       0  17.0    61.0   135.0   Yes    0.411415   
1207   87      247.0       1  18.0    66.0   125.0   Yes    1.771888   
1217  146      319.0       0  28.0    66.0   145.0    No    1.448484   
1219  110      321.0       0  28.0    66.0   180.0    No    0.516489   
1226  109      244.0       1  21.0    63.0   102.0   Yes    0.571071   

      zscore_gestation  
6             2.193357  
7             2.130871  
10            4.492686  
59            2.943194  
63            2.130871  
...                ...  
1206          2.318330  
1207          2.005898  
1217          2.493122  
1219          2.618095  
1226          2.193357  

[71 rows x 9 columns]

In [70]:
#6-3
baby['parity'].quantile(.25)

0.0

In [71]:
baby['parity'].quantile(.75)

1.0

In [72]:
q1 = baby['parity'].quantile(.25)
q3 = baby['parity'].quantile(.75)
print("Q1:", q1)
print("Q3:", q3)

Q1: 0.0
Q3: 1.0


In [70]:
baby

bwt  gestation  parity   age  height  weight smoke  zscore_bwt  \
0     120      284.0       0  27.0    62.0   100.0    No    0.029337   
1     113      282.0       0  33.0    64.0   135.0    No    0.352741   
2     128      279.0       0  28.0    64.0    77.0   Yes    0.465998   
4     108      282.0       0  23.0    67.0   125.0   Yes    0.625654   
5     136      286.0       0  25.0    62.0    93.0    No    0.902658   
...   ...        ...     ...   ...     ...     ...   ...         ...   
1231  113      275.0       1  27.0    60.0   100.0    No    0.352741   
1232  128      265.0       0  24.0    67.0   120.0    No    0.465998   
1233  130      291.0       0  30.0    65.0    67.0   Yes    0.575163   
1234  125      281.0       1  21.0    65.0   110.0    No    0.302250   
1235  117      297.0       0  38.0    65.0   129.0    No    0.134411   

      zscore_gestation  
0             0.306098  
1             0.181125  
2             0.006334  
4             0.181125  
5             0.431071  
...                ...  
1231          0.256279  
1232          0.881143  
1233          0.743503  
1234          0.118639  
1235          1.118421  

[1174 rows x 9 columns]

In [73]:
iqr = q3 - q1

print("IQR:", iqr)

IQR: 1.0


In [74]:
top = q3 + (iqr * 1.5)
bottom = q1 - (iqr * 1.5)


print("Upper Limit:", top)
print("Lower Limit:", bottom)

Upper Limit: 2.5
Lower Limit: -1.5


In [75]:
baby["zscore_parity"] = np.abs(stats.zscore(baby["parity"]))

In [76]:
print("INDEX VALUES:", iqr_outliers)

INDEX VALUES: Int64Index([   6,    7,   10,   59,   63,   66,  119,  129,  155,  188,  192,
             198,  210,  215,  217,  234,  240,  253,  260,  279,  345,  361,
             373,  394,  440,  460,  462,  484,  500,  511,  523,  630,  685,
             699,  710,  726,  746,  761,  769,  778,  784,  828,  829,  830,
             833,  869,  912,  927,  952,  969,  978, 1002, 1020, 1026, 1035,
            1065, 1074, 1134, 1139, 1140, 1142, 1146, 1152, 1172, 1178, 1199,
            1206, 1207, 1217, 1219, 1226],
           dtype='int64')


In [77]:
baby.loc[[6,    7,   10,   59,   63,   66,  119,  129,  155,  188,  192,
             198,  210,  215,  217,  234,  240,  253,  260,  279,  345,  361,
             373,  394,  440,  460,  462,  484,  500,  511,  523,  630,  685,
             699,  710,  726,  746,  761,  769,  778,  784,  828,  829,  830,
             833,  869,  912,  927,  952,  969,  978, 1002, 1020, 1026, 1035,
            1065, 1074, 1134, 1139, 1140, 1142, 1146, 1152, 1172, 1178, 1199,
            1206, 1207, 1217, 1219, 1226]]

bwt  gestation  parity   age  height  weight smoke  zscore_bwt  \
6     138      244.0       0  33.0    62.0   178.0    No    1.011823   
7     132      245.0       0  23.0    65.0   140.0    No    0.684328   
10    140      351.0       0  27.0    68.0   120.0    No    1.120989   
59     75      232.0       0  33.0    61.0   110.0    No    2.426879   
63    101      245.0       0  23.0    63.0   130.0   Yes    1.007732   
...   ...        ...     ...   ...     ...     ...   ...         ...   
1206  127      242.0       0  17.0    61.0   135.0   Yes    0.411415   
1207   87      247.0       1  18.0    66.0   125.0   Yes    1.771888   
1217  146      319.0       0  28.0    66.0   145.0    No    1.448484   
1219  110      321.0       0  28.0    66.0   180.0    No    0.516489   
1226  109      244.0       1  21.0    63.0   102.0   Yes    0.571071   

      zscore_gestation  zscore_parity  
6             2.193357       0.596371  
7             2.130871       0.596371  
10            4.492686       0.596371  
59            2.943194       0.596371  
63            2.130871       0.596371  
...                ...            ...  
1206          2.318330       0.596371  
1207          2.005898       1.676809  
1217          2.493122       0.596371  
1219          2.618095       0.596371  
1226          2.193357       1.676809  

[71 rows x 10 columns]

In [78]:
#6-4
baby['age'].quantile(.25)

23.0

In [79]:
baby['age'].quantile(.75)

31.0

In [80]:
q1 = baby['age'].quantile(.25)
q3 = baby['age'].quantile(.75)
print("Q1:", q1)
print("Q3:", q3)

Q1: 23.0
Q3: 31.0


In [81]:
baby

bwt  gestation  parity   age  height  weight smoke  zscore_bwt  \
0     120      284.0       0  27.0    62.0   100.0    No    0.029337   
1     113      282.0       0  33.0    64.0   135.0    No    0.352741   
2     128      279.0       0  28.0    64.0    77.0   Yes    0.465998   
4     108      282.0       0  23.0    67.0   125.0   Yes    0.625654   
5     136      286.0       0  25.0    62.0    93.0    No    0.902658   
...   ...        ...     ...   ...     ...     ...   ...         ...   
1231  113      275.0       1  27.0    60.0   100.0    No    0.352741   
1232  128      265.0       0  24.0    67.0   120.0    No    0.465998   
1233  130      291.0       0  30.0    65.0    67.0   Yes    0.575163   
1234  125      281.0       1  21.0    65.0   110.0    No    0.302250   
1235  117      297.0       0  38.0    65.0   129.0    No    0.134411   

      zscore_gestation  zscore_parity  
0             0.306098       0.596371  
1             0.181125       0.596371  
2             0.006334       0.596371  
4             0.181125       0.596371  
5             0.431071       0.596371  
...                ...            ...  
1231          0.256279       1.676809  
1232          0.881143       0.596371  
1233          0.743503       0.596371  
1234          0.118639       1.676809  
1235          1.118421       0.596371  

[1174 rows x 10 columns]

In [82]:
iqr = q3 - q1

print("IQR:", iqr)

IQR: 8.0


In [83]:
top = q3 + (iqr * 1.5)
bottom = q1 - (iqr * 1.5)


print("Upper Limit:", top)
print("Lower Limit:", bottom)

Upper Limit: 43.0
Lower Limit: 11.0


In [84]:
baby["zscore_age"] = np.abs(stats.zscore(baby["age"]))

In [85]:
print("INDEX VALUES:", iqr_outliers)

INDEX VALUES: Int64Index([   6,    7,   10,   59,   63,   66,  119,  129,  155,  188,  192,
             198,  210,  215,  217,  234,  240,  253,  260,  279,  345,  361,
             373,  394,  440,  460,  462,  484,  500,  511,  523,  630,  685,
             699,  710,  726,  746,  761,  769,  778,  784,  828,  829,  830,
             833,  869,  912,  927,  952,  969,  978, 1002, 1020, 1026, 1035,
            1065, 1074, 1134, 1139, 1140, 1142, 1146, 1152, 1172, 1178, 1199,
            1206, 1207, 1217, 1219, 1226],
           dtype='int64')


In [86]:
baby.loc[[6,    7,   10,   59,   63,   66,  119,  129,  155,  188,  192,
             198,  210,  215,  217,  234,  240,  253,  260,  279,  345,  361,
             373,  394,  440,  460,  462,  484,  500,  511,  523,  630,  685,
             699,  710,  726,  746,  761,  769,  778,  784,  828,  829,  830,
             833,  869,  912,  927,  952,  969,  978, 1002, 1020, 1026, 1035,
            1065, 1074, 1134, 1139, 1140, 1142, 1146, 1152, 1172, 1178, 1199,
            1206, 1207, 1217, 1219, 1226]]

bwt  gestation  parity   age  height  weight smoke  zscore_bwt  \
6     138      244.0       0  33.0    62.0   178.0    No    1.011823   
7     132      245.0       0  23.0    65.0   140.0    No    0.684328   
10    140      351.0       0  27.0    68.0   120.0    No    1.120989   
59     75      232.0       0  33.0    61.0   110.0    No    2.426879   
63    101      245.0       0  23.0    63.0   130.0   Yes    1.007732   
...   ...        ...     ...   ...     ...     ...   ...         ...   
1206  127      242.0       0  17.0    61.0   135.0   Yes    0.411415   
1207   87      247.0       1  18.0    66.0   125.0   Yes    1.771888   
1217  146      319.0       0  28.0    66.0   145.0    No    1.448484   
1219  110      321.0       0  28.0    66.0   180.0    No    0.516489   
1226  109      244.0       1  21.0    63.0   102.0   Yes    0.571071   

      zscore_gestation  zscore_parity  zscore_age  
6             2.193357       0.596371    0.992496  
7             2.130871       0.596371    0.727088  
10            4.492686       0.596371    0.039255  
59            2.943194       0.596371    0.992496  
63            2.130871       0.596371    0.727088  
...                ...            ...         ...  
1206          2.318330       0.596371    1.758838  
1207          2.005898       1.676809    1.586880  
1217          2.493122       0.596371    0.132704  
1219          2.618095       0.596371    0.132704  
1226          2.193357       1.676809    1.071005  

[71 rows x 11 columns]

In [87]:
#6-5
baby['height'].quantile(.25)

62.0

In [88]:
baby['height'].quantile(.75)

66.0

In [89]:
q1 = baby['height'].quantile(.25)
q3 = baby['height'].quantile(.75)
print("Q1:", q1)
print("Q3:", q3)

Q1: 62.0
Q3: 66.0


In [90]:
baby

bwt  gestation  parity   age  height  weight smoke  zscore_bwt  \
0     120      284.0       0  27.0    62.0   100.0    No    0.029337   
1     113      282.0       0  33.0    64.0   135.0    No    0.352741   
2     128      279.0       0  28.0    64.0    77.0   Yes    0.465998   
4     108      282.0       0  23.0    67.0   125.0   Yes    0.625654   
5     136      286.0       0  25.0    62.0    93.0    No    0.902658   
...   ...        ...     ...   ...     ...     ...   ...         ...   
1231  113      275.0       1  27.0    60.0   100.0    No    0.352741   
1232  128      265.0       0  24.0    67.0   120.0    No    0.465998   
1233  130      291.0       0  30.0    65.0    67.0   Yes    0.575163   
1234  125      281.0       1  21.0    65.0   110.0    No    0.302250   
1235  117      297.0       0  38.0    65.0   129.0    No    0.134411   

      zscore_gestation  zscore_parity  zscore_age  
0             0.306098       0.596371    0.039255  
1             0.181125       0.596371    0.992496  
2             0.006334       0.596371    0.132704  
4             0.181125       0.596371    0.727088  
5             0.431071       0.596371    0.383171  
...                ...            ...         ...  
1231          0.256279       1.676809    0.039255  
1232          0.881143       0.596371    0.555130  
1233          0.743503       0.596371    0.476621  
1234          0.118639       1.676809    1.071005  
1235          1.118421       0.596371    1.852288  

[1174 rows x 11 columns]

In [91]:
iqr = q3 - q1

print("IQR:", iqr)

IQR: 4.0


In [92]:
top = q3 + (iqr * 1.5)
bottom = q1 - (iqr * 1.5)


print("Upper Limit:", top)
print("Lower Limit:", bottom)


Upper Limit: 72.0
Lower Limit: 56.0


In [93]:
print("INDEX VALUES:", iqr_outliers)

INDEX VALUES: Int64Index([   6,    7,   10,   59,   63,   66,  119,  129,  155,  188,  192,
             198,  210,  215,  217,  234,  240,  253,  260,  279,  345,  361,
             373,  394,  440,  460,  462,  484,  500,  511,  523,  630,  685,
             699,  710,  726,  746,  761,  769,  778,  784,  828,  829,  830,
             833,  869,  912,  927,  952,  969,  978, 1002, 1020, 1026, 1035,
            1065, 1074, 1134, 1139, 1140, 1142, 1146, 1152, 1172, 1178, 1199,
            1206, 1207, 1217, 1219, 1226],
           dtype='int64')


In [94]:
baby.loc[[6,    7,   10,   59,   63,   66,  119,  129,  155,  188,  192,
             198,  210,  215,  217,  234,  240,  253,  260,  279,  345,  361,
             373,  394,  440,  460,  462,  484,  500,  511,  523,  630,  685,
             699,  710,  726,  746,  761,  769,  778,  784,  828,  829,  830,
             833,  869,  912,  927,  952,  969,  978, 1002, 1020, 1026, 1035,
            1065, 1074, 1134, 1139, 1140, 1142, 1146, 1152, 1172, 1178, 1199,
            1206, 1207, 1217, 1219, 1226]]

bwt  gestation  parity   age  height  weight smoke  zscore_bwt  \
6     138      244.0       0  33.0    62.0   178.0    No    1.011823   
7     132      245.0       0  23.0    65.0   140.0    No    0.684328   
10    140      351.0       0  27.0    68.0   120.0    No    1.120989   
59     75      232.0       0  33.0    61.0   110.0    No    2.426879   
63    101      245.0       0  23.0    63.0   130.0   Yes    1.007732   
...   ...        ...     ...   ...     ...     ...   ...         ...   
1206  127      242.0       0  17.0    61.0   135.0   Yes    0.411415   
1207   87      247.0       1  18.0    66.0   125.0   Yes    1.771888   
1217  146      319.0       0  28.0    66.0   145.0    No    1.448484   
1219  110      321.0       0  28.0    66.0   180.0    No    0.516489   
1226  109      244.0       1  21.0    63.0   102.0   Yes    0.571071   

      zscore_gestation  zscore_parity  zscore_age  
6             2.193357       0.596371    0.992496  
7             2.130871       0.596371    0.727088  
10            4.492686       0.596371    0.039255  
59            2.943194       0.596371    0.992496  
63            2.130871       0.596371    0.727088  
...                ...            ...         ...  
1206          2.318330       0.596371    1.758838  
1207          2.005898       1.676809    1.586880  
1217          2.493122       0.596371    0.132704  
1219          2.618095       0.596371    0.132704  
1226          2.193357       1.676809    1.071005  

[71 rows x 11 columns]

7. Describe the characteristics of your qualitative variables by doing the following:

    * Determine the frequencies of each group within each categorical variable. Which groups have the highest frequency?
    * Determine the relative frequencies of each group within each categorical variable. Which groups have the highest relative frequencies?
    * Create a crosstab table using both the categorical variables in your dataset. How many mothers are smokers who have been pregnant before?

In [95]:
#7-1 The "No" have the highest frequency.
baby["smoke"].value_counts()

No     715
Yes    459
Name: smoke, dtype: int64

In [117]:
baby["smoke"].value_counts(normalize=True) baby["bwt"].mode()no"

No     0.609029
Yes    0.390971
Name: smoke, dtype: float64

In [96]:
baby["parity"].value_counts() # For parity the zero gave tge highest frequency.

0    866
1    308
Name: parity, dtype: int64

In [99]:
baby["parity"].value_counts(normalize=True) # the zero have the highest relative frequency.

0    0.737649
1    0.262351
Name: parity, dtype: float64

In [105]:
pd.crosstab(baby["parity"], baby["smoke"], margins=False, normalize=False)

smoke    No  Yes
parity          
0       525  341
1       190  118

In [106]:

pd.crosstab(baby["smoke"], baby["parity"], margins=False, normalize=False) # 341 mothers
are smokers.

parity    0    1
smoke           
No      525  190
Yes     341  118

8. Describe the Measures of Central Tendency (mean, median, mode) of the numeric variables in the dataset. Calculate the mean, median, and mode for all 5 numeric variables. For each median and mode, describe the following:

    * What does the median value tell you about the data?
    * What does the mode value tell you about the data?

In [109]:
baby.describe() #the medium is the average values within in the data. For example, the age
for pregnant women is 27 years old. The mode is most frequent number - 23 year old preganant 
women show frequently shows up and the weight that frequently shows up is 130 lbs.

bwt    gestation       parity          age       height  \
count  1174.000000  1174.000000  1174.000000  1174.000000  1174.000000   
mean    119.462521   279.101363     0.262351    27.228279    64.049404   
std      18.328671    16.010305     0.440100     5.817839     2.526102   
min      55.000000   148.000000     0.000000    15.000000    53.000000   
25%     108.000000   272.000000     0.000000    23.000000    62.000000   
50%     120.000000   280.000000     0.000000    26.000000    64.000000   
75%     131.000000   288.000000     1.000000    31.000000    66.000000   
max     176.000000   353.000000     1.000000    45.000000    72.000000   

            weight   zscore_bwt  zscore_gestation  zscore_parity   zscore_age  
count  1174.000000  1174.000000       1174.000000    1174.000000  1174.000000  
mean    131.326235     0.777491          0.695097       0.879825     0.814227  
std      54.103828     0.629162          0.719222       0.475501     0.580794  
min      27.000000     0.025246          0.006334       0.596371     0.039255  
25%     114.000000     0.298159          0.243612       0.596371     0.383171  
50%     125.000000     0.625654          0.499891       0.596371     0.727088  
75%     139.750000     1.116897          0.868476       1.676809     1.164454  
max    1500.000000     3.518531          8.192051       1.676809     3.055996

In [111]:
baby["bwt"].mode()

0    115
Name: bwt, dtype: int64

In [112]:
baby["parity"].mode()

0    0
Name: parity, dtype: int64

In [113]:
baby["gestation"].mode()

0    282.0
Name: gestation, dtype: float64

In [114]:
baby["age"].mode()

0    23.0
Name: age, dtype: float64

In [115]:
baby["height"].mode()

0    64.0
Name: height, dtype: float64

In [116]:
baby["weight"].mode()

0    130.0
Name: weight, dtype: float64

9. Describe the Measures of Variability of the numeric variables in the dataset. Calculate the range, standard deviation, and 85th percentile for all 5 numeric variables. Answer the following questions:

    * Which column has the largest range? Which has the smallest? What does this mean?
    * Which column has the largest standard deviation? Which has the smallest? What does this tell you about those variables?

In [ ]:
#Weight has the highest range (1473) and height has the lowest range (19)

# Age has the lowest std deviation 5.81 and weight has the highest standard deviation 54.4 
# which means both are outliers.

In [98]:
#9-1
hrs_range = baby['bwt'].max() - baby['bwt'].min()

print(hrs_range)

121


In [99]:
print("The birth weight of the baby:", baby["bwt"].mean())

baby["bwt"].std()

The birth weight of the baby: 119.46252129471891


18.32867144332098

In [100]:
np.percentile(baby['bwt'], 50)

120.0

In [114]:
hrs_range = baby['gestation'].max() - baby['gestation'].min()

print(hrs_range)

205.0


In [103]:

print("9 months of gestation:", baby["gestation"].mean())

baby["gestation"].std()

9 months of gestation: 279.1013628620102


16.010305127129392

In [104]:
np.percentile(baby['gestation'], 50)

280.0

In [105]:
hrs_range = baby['age'].max() - baby['age'].min()

print(hrs_range)

30.0


In [106]:
print("The age of the mother while pregnant:", baby["age"].mean())


The age of the mother while pregnant: 27.228279386712096


In [115]:
baby["age"].std()

5.817838716855792

In [107]:
np.percentile(baby['age'], 50)

26.0

In [108]:
hrs_range = baby['height'].max() - baby['height'].min()

print(hrs_range)

19.0


In [109]:
print("What is the mean of the babies?:", baby["height"].mean())


What is the mean of the babies?: 64.04940374787053


In [116]:
baby["height"].std()

2.5261015226020365

In [110]:
np.percentile(baby['height'], 50)

64.0

In [112]:
hrs_range = baby['weight'].max() - baby['weight'].min()

print(hrs_range)

1473.0


In [113]:
print("What is the mean of the babies' weight?:", baby["weight"].mean())


What is the mean of the babies' weight?: 131.32623509369677


In [117]:
baby["weight"].std()

54.1038284756229

In [118]:
np.percentile(baby['weight'], 50)

125.0